kahypar/partition/bin_packing/bin_packer.h

```diff
@@ -0,0 +1,124 @@
+
+        // The performance of the prepacking algorithm depends on the perfect part weight, the allowed weight for each bin
+        // and the allowed part weight, which must be between those, i.e. lower <= border <= upper. If border has a value close
+        // to lower, the possibilities of the initial partitioning algorithm are rather restricted, but only few vertices need
+        // to be prepacked. On the other hand, if border has a value close to upper, almost every vertex will be prepacked,
+        // resulting in a bisection with a poor cut.
+        // To avoid this, we check for this case and adjust the allowed part weight if necessary.
+        if (upper - border < (border - lower) / 10) {
+          border = (lower + upper) / 2;
+          _context.partition.epsilon = static_cast<double>(border) / static_cast<double>(lower) - 1.0;
+        }
+      }
+
+      bin_packing::calculateExactPrepacking<BPAlgorithm>(_hypergraph, _context, rb_range_k, max_bin_weight);
+    }
+  }
+
+  std::vector<PartitionID> twoLevelPackingImpl(const std::vector<HypernodeID>& nodes, const HypernodeWeight max_bin_weight) const override {
+    ASSERT(static_cast<size_t>(_context.partition.k) == _context.initial_partitioning.upper_allowed_partition_weight.size());
+
+    const PartitionID rb_range_k = _context.partition.rb_upper_k - _context.partition.rb_lower_k + 1;
+    std::vector<PartitionID> parts(nodes.size(), -1);
+    TwoLevelPacker<BPAlgorithm> packer(rb_range_k, max_bin_weight);
+
+    if (_hypergraph.containsFixedVertices()) {
+      bin_packing::preassignFixedVertices<BPAlgorithm>(_hypergraph, nodes, parts, packer, _context.partition.k, rb_range_k);
+    }
+
+    // At the first level, a packing with rb_range_k bins is calculated ...
+    for (size_t i = 0; i < nodes.size(); ++i) {
+      const HypernodeID hn = nodes[i];
+
+      if(!_hypergraph.isFixedVertex(hn)) {
+        HypernodeWeight weight = _hypergraph.nodeWeight(hn);
+        parts[i] = packer.insertElement(weight);
+      }
+    }
+
+    // ... and at the second level, the resulting bins are packed into the final parts.
+    PartitionMapping packing_result = packer.applySecondLevel(_context.initial_partitioning.upper_allowed_partition_weight,
+                                                              _context.initial_partitioning.num_bins_per_part).first;
+    packing_result.applyMapping(parts);
+
+    ASSERT(nodes.size() == parts.size());
+    ASSERT([&]() {
+      for (size_t i = 0; i < nodes.size(); ++i) {
+        const HypernodeID hn = nodes[i];
+
+        if (_hypergraph.isFixedVertex(hn) && (_hypergraph.fixedVertexPartID(hn) != parts[i])) {
+          return false;
+        }
+      }
+      return true;
+    } (), "Partition of fixed vertex changed.");
+
+    return parts;
+  }
+
+  Hypergraph& _hypergraph;
+  Context& _context;
+};
+
+static std::unique_ptr<IBinPacker> createBinPacker(const BinPackingAlgorithm& bp_algo, Hypergraph& hypergraph, Context& context) {
```

Comment:
For consistency, I would propose to use our `Factory` approach here (see: `kahypar/partition/factories.h`). @kittobi1992 can help you with that, if you run into any problems.